packages/ckeditor5-paragraph/src/insertparagraphcommand.ts

```diff
@@ -64,15 +64,34 @@ export default class InsertParagraphCommand extends Command {
 				model.schema.setAllowedAttributes( paragraph, attributes, writer );
 			}
 
-			if ( !model.schema.checkChild( position.parent as Element, paragraph ) ) {
-				const allowedParent = model.schema.findAllowedParent( position, paragraph );
+			const allowedParent = model.schema.findAllowedParent( position, paragraph );
 
-				// It could be there's no ancestor limit that would allow paragraph.
-				// In theory, "paragraph" could be disallowed even in the "$root".
-				if ( !allowedParent ) {
-					return;
-				}
+			// It could be there's no ancestor limit that would allow paragraph.
+			// In theory, "paragraph" could be disallowed even in the "$root".
+			if ( !allowedParent ) {
+				return;
+			}
 
+			if ( position.isAtStart && position.isAtEnd && position.path.length > 1 ) {
+				// NOTE: In case when paragraph is empty.
+				// <paragraph>[]</paragraph> ---> <paragraph></paragraph><paragraph>[]</paragraph>
+				position = createPositionBellow( writer, position, allowedParent );
+			} else if ( position.isAtStart && position.path.length > 1 && ( position.nodeAfter as Text ).data ) {
+				// When position is at the start of the line we want to insert paragraph above.
+				// <paragraph>[]foo</paragraph> ---> <paragraph>[]</paragraph><paragraph>foo</paragraph>
+				position = writer.createPositionFromPath( allowedParent, position.getParentPath() );
+			} else if ( position.isAtEnd && position.path.length > 1 && ( position.nodeBefore as Text ).data ) {
```

Comment:
All these conditions contain `position.path.length > 1`. Maybe it's possible to simplify them? What does it mean if it's equal to 0 or 1? Is an earlier condition possible?

Reply:
If **path** contain only 1 value like [ 1 ], `getParentPath()` of that will be empty array ([ ]) then it's not valid path.

Reply:
Eventually we could replace it with
`
if ( position.path.length < 2 ) {
				model.insertContent( paragraph, position );
				writer.setSelection( paragraph, 'in' );

				return;
			}

			if ( position.isAtStart && position.isAtEnd ) {
				// NOTE: In case when paragraph is empty.
				// <paragraph>[]</paragraph> ---> <paragraph></paragraph><paragraph>[]</paragraph>
				position = writer.createPositionAfter( position.parent as Item );
			} else if ( position.isAtStart && ( position.nodeAfter as Text ).data ) {
				// When position is at the start of the line we want to insert paragraph above.
				// <paragraph>[]foo</paragraph> ---> <paragraph>[]</paragraph><paragraph>foo</paragraph>
				position = writer.createPositionFromPath( allowedParent, position.getParentPath() );
			} else if ( position.isAtEnd && ( position.nodeBefore as Text ).data ) {
				// When position is at the end of the line we want to insert paragraph after.
				// <paragraph>foo[]</paragraph> ---> <paragraph>foo</paragraph><paragraph>[]</paragraph>
				position = writer.createPositionAfter( position.parent as Item );
			} else if ( !model.schema.checkChild( position.parent as Element, paragraph ) ) {
				// When position is inside the content we want to split it
				// <paragraph>fo[]o</paragraph>
				//            |
				//			  ↓
				// <paragraph>fo</paragraph>
				// <paragraph>[]</paragraph>
				// <paragraph>o</paragraph>
				position = writer.split( position, allowedParent ).position;
			}

			model.insertContent( paragraph, position );

			writer.setSelection( paragraph, 'in' );
`